Add notes for Pittsburgh CG meeting (#1629)
WebAssembly/meetings: main/2024/CG-06.md

- 2:00pm: JS string builtins (Ryan Hunt)
    - Vote to advance to phase 3
- 2:45pm: Break

> - 3:00pm: Memory control (Deepti Gandluri)

Reviewer: I think these agenda diffs are reverts of the agenda changes made after this file landed in the repo but before the meeting.

Author: Thanks, fixed. I left off the link to Luke's slides because it appears again alongside the notes for that session.

BT: I found a website that lists all the engines, and I sent a PR to the website, but it took three months to land. We should have a solution where our burden is low and engine authors' burden is clear. We should reward the folks working on the engines, or other pages with more traffic so our stuff doesn’t get out of date and their stuff also gets visibility. 

SE: I like the idea of having organizations committing to donating resources for this. Seems like we could accept donations when organizations make them, and do minimal maintenance otherwise.

AK: The immediate problem is that the website is very out of date. If we have an order of operations, start with surveying what's there, and update the documentation, and then come back to adding new features. We’ll then have a structure in place that is responsible for maintenance as well.

David Thompson (chat): a simple static site with good docs is all that is really needed, in my opinion. The feature matrix is particularly important to me. I often forget which version of which browser supports which proposal.

#### Custom annotations (Andreas Rossberg)

> Andreas Rossberg Presenting Custom annotations (TODO: Add slides)

Reviewer: ```suggestion
Andreas Rossberg Presenting Custom annotations [slides](/2024-06-05-rossberg-annotations.pdf)
```

> Andreas Rossberg Presenting Custom annotations [slides](/2024-06-05-rossberg-annotations.pdf)

Author: ```suggestion
Andreas Rossberg Presenting Custom annotations [slides](presentations/2024-06-05-rossberg-annotations.pdf)
```

DD: That’s clear. For names, it’s good that the standard is calling it out explicitly. On the tooling side, I think it's good to announce major non-standard custom sections that people are going to use.

AR: In terms of actual custom custom sections that tools use, it would be nice to have some convention to pick names for them to avoid name clobbering. We as a community could define these conventions, but that is orthogonal to the custom sections problem itself.

Dave Thompson (chat): So would the name section be populated by "name" annotations rather than the existing method of extracting names from the symbolic ids of various things (types, funcs, etc.)?

AR: How you construct the Wasm module is up to the tools, there’s no prescriptive ways on how to do it.

**POLL**: Custom annotations to Phase 4

|SF|F|N|A|SA|
|--|-|-|-|--|
|14|30|8|0|0|

TL: That is consensus for Phase 4.

Room: *applause*

#### Branch hinting (Yuri Iozzelli)

> YI presenting branch hinting (TODO: Add slides)

Reviewer: @yuri91 do you have a slides link?

Reviewer 2: https://drive.google.com/file/d/12ojxNtRrugs-4MrLqOghCCn2fJbfmqcl/view?usp=sharing

> YI presenting branch hinting (TODO: Add slides)

Author: ```suggestion
YI presenting branch hinting [slides](https://drive.google.com/file/d/12ojxNtRrugs-4MrLqOghCCn2fJbfmqcl/view)
```

AR: My take would be that we could go a similar route with these that we are trying with string builtins. Particularly once we have something like type imports. Right now we use externref, but we could have interop modules in the JS API level that give you the primitives you need.

SD: Yes that would work.

CW: To echo Andreas, the direction I expect this to go would be to facilitate the allocation of JS objects instead of trying to make engines more clever behind the scenes.

AR: One other comment re: closures: The main reason we removed them was concerns that we want a plain function pointer type, and don’t want to conflate it with a closure type. There might be room for closures as a future feature but would be a separate type. They are a richer thing with more expansive implementation.

SD: Sure, that wouldn't fundamentally change how we use it.

AR: With all of these things, it really takes someone to champion it and push it through. Now it seems that we do have a user.

CW: To be fair, the rebuttal is often to have the engines implement these proposals as that often takes significant work.

AR: Yes that too. 

TL: Thank you very much!

#### JS string builtins (Ryan Hunt)

> RH presenting. [TODO: slides]

Reviewer: @eqrion do you have a slides link?

Author: Friendly bump @eqrion

Reviewer 3: https://docs.google.com/presentation/d/111gkNGmzOtATU7daMKBjZ6mBqbIXF6jhd3e1542-0xo/edit#slide=id.g1ef40614c6f_0_21

Sorry, for the delay.

BT: Why not the empty string?

RH: Yeah I like that.

CW: Now I’m thinking about this new global trick…it’s interesting because previously we could understand it as a special case of compile time imports, now it’s compile time imports + import name reflection. One concrete question: for normal wasm modules, the import things will be quite small, so are there implementation defined limits for import names that will clash with these string constants?

RH: I’m sure we have one, feels familiar. But haven’t heard from anyone experimenting with it that it’s been an issue, and it’s something we could easily increase.

BT: Just to prompt your imagination a bit. Instead of a quote, just a json literal. The engine has amazing optimization for json and would create a literal expediently.

RH: I will file an issue, you’re all invited to comment. Phase 3 poll?

TL: Poll to advance to phase 3!

|SF|F|N|A|SA|
|--|-|-|-|--|
|17|14|12|0|0

TL: That is consensus.

Room: *applause*

#### Memory control (Deepti Gandluri and Ben Visness)

> DG & BV presenting slides (TODO: link to slides); notes not already in slides:

Reviewer: @dtig @bvisness do you have a slide link?

Reviewer 4: Here: https://docs.google.com/presentation/d/1n89xOMpZMpic9MBIuR3mkxABQCsYG6U7lr33TdWlVpQ/edit?usp=sharing

> DG & BV presenting slides (TODO: link to slides); notes not already in slides:

Author: ```suggestion
DG & BV presenting [slides](https://docs.google.com/presentation/d/1n89xOMpZMpic9MBIuR3mkxABQCsYG6U7lr33TdWlVpQ/edit?usp=sharing); notes not already in slides:
```

DG: Since here it’s not a web-only feature I’m not sure we should gate on web tests.

BT: We could add JS API tests in the spec repo.

DS: I had trouble running the JS API tests.

BT: Does it even say the tests must pass?

(It does)

BT: Let’s defer the vote until we have the tests.

HA: I really think we can have them done in a week, but might not be merged to WPT repo. I’m happy to come back in a couple of weeks.

AK: We’ve done votes assuming something will happen…

BT: Lets not do that again.

TL: Let’s clarify the JS tests.

AR: It matters where it is and it should be in the spec repo.

HA: Is there a way to run them without WPT?

AR: Uou can set a path to node and it will run them.

DG: You can import spec tests to V8 to run them. This is a bit unique because we changed the spec at phase 3.

TS: Node and deno run relevant WPT tests. I think they should be in the spec

CW: What’s the fastest place to put tests? The spec repo.

RH: I'm pretty sure we can run tests from the spec repo.

AK: I’m happy to help.

BT: All agree.

HA: What's our conclusion?

BT: Put tests in the spec repo and then vote on phase 4.
### Thursday, June 6
#### SpecTec DSL (Andreas Rossberg)

> AR Presenting [TODO: link to slides]

Reviewer: ```suggestion
AR Presenting [slides](presentations//2024-06-06-rossberg-spectec.pdf)
```

Deepti: You had talked about a process for fallbacks. Should requiring the tool be extended to meet phase 4 requirements for proposals be required? Should we consider the process on a proposal-by-proposal basis? Overall, I think there are a lot of process details to figure out.

AR: In terms of infrastructure there isn’t so much needed. In our fork, CI runs the tool and generates the document. We’d need to add some config to disable individual backends. I’m not too worried about that part. For the process part, the question is, who is responsible for updating SpecTec to meet the needs of prospective phase 4 proposals, and on what timeline? These are questions the CG needs to consider.

DS: Let’s go ahead and do the vote.

Poll: Adopt SpecTec (once it’s ready) as the toolchain for authoring the spec.

SF: 19 in room, 5 on chat
F: 13 in room, 4 online
N: 4 in room, 4 online
A: 1 in room
SA: 0

DS: Consensus.
#### Typed Continuations (Frank Emrich)

> FE presenting [TODO: Add slides]

Reviewer: @frank-emrich do you have a link to the slides?

> FE presenting [TODO: Add slides]

Author: ```suggestion
FE presenting [slides](presentations/2024-06-06-emrich-wasmfx-stack-switching.pdf)
```

AR: Disagree: a loop is just a label for a backedge.

Mattias: It forces structured control flow.

AR: Yes, it restricts things, but in terms of semantics it’s not a particularly weird feature.

Mattias: That was maybe a bad example. Trying to convey the idea that we’re trying to be high level and low level can be bad. If we try to go too low level, we make a choice at compile time.

AR: I have often used the slogan: “as low level as possible but no lower”. Of course open to interpretation, but definitely things where we are higher level than a CPU, and a function is a clear example. Sometimes we have to go to a high level, but usually want to stay as low as you possibly can.

Petr: Can we go on a deeper dive on this, maybe in CG meeting?

DS: Yes.

Petr: Yes, would like to discuss tools and what solutions are.
#### Large Wasm binaries from desktop application migrations (David Sankel)

> DVS presenting. (I am referring to him as DVS to be distinct from Derek.) [TODO: Slides]

Reviewer: @camio do you want to publish your slides?

> DVS presenting. (I am referring to him as DVS to be distinct from Derek.) [TODO: Slides]

Author: ```suggestion
DVS presenting [slides](presentations/2024-06-06-sankel-big-wasm-binaries.pdf). (I am referring to him as DVS to be distinct from Derek.)
```

CW: RH, can you talk about thread local globals?

RH: Have discussed that a lot in issues, and we don’t have a lot of time right now, so redirecting with a new question: What are your thoughts on how to proceed on the issue? 

TL: Yes, we’ve had a lot of discussions and they’ve been productive, and they’ve stopped. So we are in a position to wait until we have a prototype and some performance numbers.

RH: Are you interested in implementing a different approach so that you can compare them? Especially if the first implementation is obviously slow?

TL: If the first implementation is fast, maybe that’s good enough?

LW: numbers will help with “can tls be fast?” it won’t help with the shared-to-unshared edges discussion.

TL, RH: Agreed.

RH: It’s not just a big project in SpiderMonkey, but also in Gecko, to get right, and that’s concerning.

TL: We need to wait for more implementor feedback. What do they need, what can they get away with? Evaluate the tradeoff in light of that data.

RH: Toolchains are going to want the stronger semantics.

TL: We can push back and ask for feedback;“what if you can’t have the stronger semantics?”
#### Memory64 (Sam Clegg)

> TODO: link to slides slides 

Reviewer: @sbc100 do you have a link to your slides?

Author: Friendly bump @sbc100

Reviewer 5: Sorry, I'm was sure I posted this already: https://docs.google.com/presentation/d/1LRIIOUHZoTm127fnM-BuUDjVLOgUUzvqISJgiI_fWeY/edit?usp=sharing
